### ISSUE TITLE
Way to specify VolumeSnapshotClass while taking a VMSnapshot if this is not there already 

### ISSUE THREAD

**Issue by socceranoo** [kubevirt/kubevirt#13199]
**Is your feature request related to a problem? Please describe**:
THere should be a way to specify VolumeSnapshotClass while creating a VMSnapshot

**Describe the solution you'd like**:

**Describe alternatives you've considered**:

**Additional context**:
There could be multiple VolumeSnapshotClasses configured on the kubernetes cluster. So there should be a way to select one while takiing a VMsnapshot

**Comment by socceranoo** [comment]
Will this handle the case where there could be multiple volumes in the VirtualMachine coming from different storage classes ? If so how do we specify that in the vmsnapshot object ?

**Comment by awels** [comment]
So VMSnapshots use volume snapshots to create the disk snapshots. Each storage class can have one or more volume snapshot classes associated with it. IIRC right now if there is only one volume snapshot class we use that, if there are multiple we use whichever one is set as the default volume snapshot class. I would have to go look up the behavior if no default volume snapshot class is set and multiple volume snapshot classes are associated with one storage class. Note you can have multiple storage classes each with their own volume snapshot classes and each of those can have a default.

**Comment by socceranoo** [comment]
i was referring to this PR https://github.com/kubevirt/kubevirt/pull/13200
if a volumesnapshotclass is added to the VMSnapshot spec, which PVC's volumesnapshotClass we are specifying here ? Since there could be multiple PVCs under the VirtualMachineSnapshotContent which is created by VirtualMachineSnapshot

Is there a mechanism where volumesnapshot objects can be created and attached to the VMsnapshot object then it would user more control at each PVC level.

**Comment by awels** [comment]
I didn't see that PR, I still don't understand the need to do this. The snapshot class is selected based on the provisioner of the PVC you are snapshotting. What is the use case of that PR? I will comment on the PR to see why they want that.

**Comment by mhenriks** [comment]
cc @ShellyKa13

**Comment by arnongilboa** [comment]
In case CDI is in the picture, we can clone the `StorageClass`, so we have two `StorageProfiles` - for local snapshots and remote backups, and we can configure each with the relevant `VolumeSnapshotClass`. See [here](https://github.com/kubevirt/containerized-data-importer/pull/2898).

**Comment by socceranoo** [comment]
Yes CDI is in picture. If i understand correctly, storage profile will come into picture during create time, So if we clone the storage class and create a storageprofile with a different snapshotclass, then all the snapshots for that DV/PVC will be using the new snapshot class.

The requirement is to be able to select the snapshotclass at the time of the virtual machine snapshot creation (based on the local snapshot or remote backup) . 
Can we pass the storage profile while taking the VirtualMachineSnapshot ?

**Comment by arnongilboa** [comment]
The `StorageProfile` has the name of the `StorageClass` so you cannot choose the one, however if authorized, you may patch (the cloned) `StorageProfile` with the desired `VolumeSnapshotClass` just before taking the `VirtualMachineSnapshot`. wdyt?

**Comment by arnongilboa** [comment]
Not sure about scaling in your scenario but you can clone the StorageClass per VM, so each will have its own StorageProfile.

**Comment by ShellyKa13** [comment]
I'm not sure if this exactly meets the requirement, but there is a PR I wrote that takes into account the Snapshotclass specified in the Storageprofile, if one exists. In such a situation, priority will be given to what is specified there instead of choosing the first VolumeSnapshotClass if there are multiple. Of course, this also takes into account the SC for each PVC.
